# Testing

The most fundamental way of testing code in JS is to `throw` an `error` when a condition is not met.

```js
function sum(a, b) {
  return a + b
}

const result = sum(2, 3)
const expected = 5

if (result !== expected) {
  throw new Error(`${result} is not equal to ${expected}`)
}

// START to optimize //////////////////////////////////////////////////////////

// Build our own expect function with a variety of assertion methods-----------
function expect(actual) {
  return {
    toBe(expected) {
      if (actual !== expected) {
        throw new Error(`${actual} is not equal to ${expected}`)
      }
    },
  }
}

// Now we can do:
expect(result).toBe(expected)

// Optimise
function test(title, callback) {
  try {
    callback()
    console.log(`✓ ${title}`)
  } catch (error) {
    console.error(title)
    console.error(`⨉ ${error}`)
  }
}

// Organizing code
function sumTest() {
  const result = sum(2, 3)
  const expected = 5
  expect(result).toBe(expected)
}
test('Sum 2 numbers', sumTest)

// Refactor
test('Sum 2 numbers', () => {
  const result = sum(2, 3)
  const expected = 5
  expect(result).toBe(expected)
})

// Test async functions -------------------------------------------------------
async function asyncTest(title, callback) {
  try {
    await callback()
    console.log(`✓ ${title}`)
  } catch (error) {
    console.error(title)
    console.error(`⨉ ${error}`)
  }
}

async function sumAsync(a, b) {
  return new Promise(res => setTimeout(res(a + b), 1000))
}

// Running sync code will work here as well
asyncTest('sumAsync adds 2 numbers asynchromously', async () => {
  const result = await sumAsync(2, 3)
  const expected = 5
  expect(result).toBe(expected)
})
```
